core/src/main/java/com/yahoo/oak/BasicChunk.java

```diff
@@ -9,55 +9,503 @@
+     * @return  true if the entry index has a valid value reference
+     *          (No check for off-heap deleted bit!)
+     */
+    boolean readValue(ValueBuffer value, int ei) {
+        if (!isIndexInBound(ei)) {
+            value.invalidate();
+            return false;
+        }
+        long reference = getValueReference(ei);
+        return value.getSlice().decodeReference(reference);
+    }
+
+    /**
+     * Reads a key from entry at the given entry index (from off-heap).
+     * Returns false if:
+     *   (1) there is no such entry or
+     *   (2) entry has no key set
+     *
+     * @param key the buffer that will contain the key
+     * @param ei  the entry index to read
+     * @return true if the entry index has a valid key allocation reference
+     */
+    boolean readKey(KeyBuffer key, int ei) {
+        if (!isIndexInBound(ei)) {
+            key.invalidate();
+            return false;
+        }
+
+        long reference = getKeyReference(ei);
+        return key.getSlice().decodeReference(reference);
+    }
+
+    protected boolean isIndexInBound(int ei) {
+        // The actual capacity is (entriesCapacity-1) because the first entry is a dummy.
+        return (ei != INVALID_ENTRY_INDEX && ei < array.entryCount());
+    }
+
+    /* ########################################################################################
+       # Methods for managing the entry context of the keys and values inside ThreadContext
+       ######################################################################################## */
+
+    /**
+     * Updates the key portion of the entry context inside {@code ctx} that matches its entry context index.
+     * Thus, {@code ctx.initEntryContext(int)} should be called prior to this method on this {@code ctx} instance.
+     *
+     * @param ctx the context that will be updated and follows the operation with this key
+     */
+    void readKey(ThreadContext ctx) {
+        readKey(ctx.key, ctx.entryIndex);
+    }
+
+    /**
+     * Updates the value portion of the entry context inside {@code ctx} that matches its entry context index.
+     * This includes both the value itself, and the value's state.
+     * Thus, {@code ctx.initEntryContext(int)} should be called prior to this method on this {@code ctx} instance.
+     *
+     * @param ctx the context that was initiated by {@code readKey(ctx, ei)}
+     */
+    void readValue(ThreadContext ctx) {
+        readValue(ctx.value, ctx.entryIndex);
+        ctx.entryState = getValueState(ctx.value);
+        assert valuesMemoryManager.isReferenceConsistent(ctx.value.getSlice().getReference());
+    }
+
+    /**
+     * Find the state of a the value that is pointed by {@code value}.
+     * Thus, {@code readValue(value, ei)} should be called prior to this method with the same {@code value} instance.
+     *
+     * @param value a buffer object that contains the value buffer
+     */
+    private EntryState getValueState(ValueBuffer value) {
```

Comment:
It is weird we have both `getEntryState()` and `getValueState()` in one file. They are doing basically the same just one for Hash's algorithm insertion/deletion and the later for Map's algorithm of insertion/deleted... Not experienced coder will get confused...

Reply:
You're right. These should be moved to the specific Chunk implementation.